[REVIEW] Refactor some of the CSV Reader kernels into generic utility functions

vuule opened the pull request:
Set up utility functions for file parsing:
* Replace the countRecords kernel with thrust library calls;
* Make the launch_countRecords into a generic utility function;
* Make the storerecordStart kernel and launch_storerecordStart generic;
* move all of the above into a separate file, making the code reusable;
Other:
* Use Async CUDA calls for memory copies and sets;
* fix chunk allocation for small files - no longer allocates full chunk size when the file is smaller than chunk size;
* Refactor the code that set the record starts into a separate function;
* Rename the CSV specific typedefs to be more generic. usable in the utility functions;

mjsamoht commented:
@OlivierNV @j-ieong can you help review this change?

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+#include "parsing_utils.cuh"
```
Missing copyright header

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+#include "rmm/thrust_rmm_allocator.h"
+#include "utilities/error_utils.hpp"
+
+constexpr size_t max_chunk_bytes = 256*1024*1024; // 256MB
```
There should be a comment on why this magic number was chosen.

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+		CUDA_TRY(cudaMemcpyAsync(d_chunk, h_chunk, chunk_bytes, cudaMemcpyDefault));
+
+		for (char key: keys) {
+			count += thrust::count(rmm::exec_policy()->on(0), d_chunk, d_chunk + chunk_bytes, key);
```
I'd rather we re-use the same kernel that we use for storing the start offsets. The logic for the paths to me is that we count in the first pass, then store in the second pass.

We re-use the same kernel in the Parquet reader; we launch to count the page headers, allocate memory, then launch to decode the page headers.

vuule replied:
Done

j-ieong commented on cpp/src/io/utilities/parsing_utils.cuh:
```diff
@@ -0,0 +1,13 @@
+
+#include "cudf.h"
+
+using ll_uint_t = unsigned long long int;
```
Why the change in type name? I'd rather we keep the higher-level logical naming like pos_offset_type or similar (like gdf_size_type). Or we be explicit in type size like uint64_t.

vuule replied:
Used the alias because CUDA atomic operations are very picky about the parameter type. However, with the recent code refactoring, this is no longer needed. Changed to uint64_t.

j-ieong commented on cpp/src/io/utilities/parsing_utils.cuh:
```diff
@@ -0,0 +1,13 @@
+#pragma once
```
Missing copyright header

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+	CUDA_TRY( cudaGetLastError() );
+
+	return GDF_SUCCESS;
+}
```
Missing newline at end of file. Some IDEs and tools require a line ending at the end

j-ieong commented on cpp/src/io/utilities/parsing_utils.cuh:
```diff
@@ -0,0 +1,13 @@
+	gdf_size_type &rec_cnt);
+
+gdf_error findAllFromSet(const char *h_data, size_t h_size, std::vector<char> keys, ll_uint_t result_offset,
+	ll_uint_t *recStart);
```
Missing newline at end of file. Some IDEs and tools require a line ending at the end

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+
+/**---------------------------------------------------------------------------*
+ * @brief For each of the characters in the input set, find and saves all 
+ * cccurences in the input host array of characters.
```
Typo in `cccurences`.

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief CUDA kernel that finds all occurences of a character in the given 
```
Typo in `occurences`

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+
+/**---------------------------------------------------------------------------*
+ * @brief Searches the input character array for each of characters in a set
+ * and sums up the number of occurences.
```
Typo in `occurences`

j-ieong commented on cpp/src/io/utilities/parsing_utils.cu:
```diff
@@ -0,0 +1,143 @@
+ * @param[in] h_data Pointer to the data in host memory
+ * @param[in] h_size Size of the input data, in bytes
+ * @param[in] keys Vector containing the keys to count in the buffer
+ * @param[out] count Total number of occurences of all keys
```
Typo in `occurences`

vuule commented:
@j-ieong Addressed all of your suggestions, please review.

Also refactored the code a bit, based on newly discovered JSON requirements, and the transition to exception-based error propagation (no return error code).
